### PR TITLE
Fix history search param _at in SQL Server

### PR DESCRIPTION
**Pull request by rbans96**
## Description
The code changes modifies the /history/_at= behavior as expected in the bug
https://github.com/microsoft/fhir-server/issues/1084

## Related issues
Addresses [#74811](https://microsofthealth.visualstudio.com/Health/_boards/board/t/Resolute/Stories/?workitem=74811)

## Testing
Unit test and e2e test are added

Here is the updated SQL query for /history/_at=[] search

DECLARE @p0 Int = 11;
DECLARE @p1 SmallInt = 96;
DECLARE @p2 VarChar(64) = '53e959f5-a007-4207-9728-1630f1bf3819';
DECLARE @p3 BigInt = 5098579070714800000;

SET STATISTICS IO ON;
SET STATISTICS TIME ON;

SELECT TOP (@p0) r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(1 AS bit) AS IsMatch, r.RawResource
FROM dbo.Resource r
WHERE ResourceTypeId = @p1
    AND ResourceId = @p2
    AND ResourceSurrogateId <= @p3
    AND NOT EXISTS( 
        SELECT *
        FROM dbo.Resource r2
        WHERE ResourceSurrogateId <= @p3
            AND r2.ResourceId = r.ResourceId
            AND r2.ResourceSurrogateId > r.ResourceSurrogateId
        )
ORDER BY r.ResourceSurrogateId ASC
OPTION(RECOMPILE)

### CODE REVIEW

**Comment by johnstairs** [comment]
I think your query for `history?_at=2020` should look like this:
```
SELECT TOP (@p0) r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(1 AS bit) AS IsMatch, r.RawResource
FROM dbo.Resource r
WHERE ResourceSurrogateId <= @p1
    AND NOT EXISTS (
        SELECT *
        FROM dbo.Resource r2
        WHERE r2.ResourceSurrogateId <= @p1
            AND r2.ResourceTypeId = r.ResourceTypeId
            AND r2.ResourceId = r.ResourceId
            AND r2.ResourceSurrogateId > r.ResourceSurrogateId
        )
ORDER BY r.ResourceSurrogateId ASC
OPTION(RECOMPILE)
```
You are missing the `r2.ResourceTypeId = r.ResourceTypeId`.

In addition to being more correct, it performs a lot faster ;)

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs [hunk 1]
```diff
@@ -120,6 +120,40 @@ public object VisitSqlRoot(SqlRootExpression expression, SearchOptions context)
...
+                    {
+                        StringBuilder.AppendLine("SELECT *");
+                        StringBuilder.Append("FROM ").Append(VLatest.Resource).Append(" ").AppendLine(resourceTableAlias2);
+                        using (var delimitedClause = StringBuilder.BeginDelimitedWhereClause())
+                        {
+                            delimitedClause.BeginDelimitedElement();
+                            foreach (var denormalizedPredicate in expression.DenormalizedExpressions)
+                            {
+                                SearchParameterExpressionBase exp = denormalizedPredicate as SearchParameterExpressionBase;
+                                if (string.Equals(exp.Parameter.Name, SqlSearchParameters.ResourceSurrogateIdParameterName, StringComparison.Ordinal))
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:140]: exp could (theoretically) be null

**Comment by rbans96** [comment]
> I think your query for `history?_at=2020` should look like this:
> 
> ```
> SELECT TOP (@p0) r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(1 AS bit) AS IsMatch, r.RawResource
> FROM dbo.Resource r
> WHERE ResourceSurrogateId <= @p1
>     AND NOT EXISTS (
>         SELECT *
>         FROM dbo.Resource r2
>         WHERE r2.ResourceSurrogateId <= @p1
>             AND r2.ResourceTypeId = r.ResourceTypeId
>             AND r2.ResourceId = r.ResourceId
>             AND r2.ResourceSurrogateId > r.ResourceSurrogateId
>         )
> ORDER BY r.ResourceSurrogateId ASC
> OPTION(RECOMPILE)
> ```
> 
> You are missing the `r2.ResourceTypeId = r.ResourceTypeId`.
> 
> In addition to being more correct, it performs a lot faster ;)

Good point, Fixed

**Comment by johnstairs** [comment]
We throw an error when the query specifies both `_at` and `_since`/`before`, but we really shouldn't. If we support both, then the current approach of put an `AtParam` on SearchOptions and reducing both to expressions on `ResourceSurrogateId` will not work, because we need to know which expression refers to the _at so that we can include the semi-join for the right predicate.

**Comment by rbans96** [comment]
> We throw an error when the query specifies both `_at` and `_since`/`before`, but we really shouldn't. If we support both, then the current approach of put an `AtParam` on SearchOptions and reducing both to expressions on `ResourceSurrogateId` will not work, because we need to know which expression refers to the _at so that we can include the semi-join for the right predicate.

I have added a flag to identify the _at expression to be able to include just that predicate in the semi-join for resourceSurrogateId. This approach seems to work for the query specifying both _at and _since parameter. Also, I will create a new user story to enable both _at and _since/_before.

**Comment by rbans96** [comment]
Closing this PR as it is considerably old and there is a plan to redesign the _at a bit and create a new PR.

**Pull request closed without merging**